var/spack/repos/builtin/packages/neurodamus-core/package.py

```diff
@@ -3,17 +3,18 @@
 from spack import *
 import os
 import shutil
-
+from spack.fetch_strategy import GitFetchStrategy
 
 class NeurodamusCore(Package):
     """Library of channels developed by Blue Brain Project, EPFL"""
 
     homepage = "ssh://bbpcode.epfl.ch/sim/neurodamus-core"
     git      = "ssh://bbpcode.epfl.ch/sim/neurodamus-core"
 
-    version('develop', git=git, branch='master')
-    version('2.3.3', git=git, tag='2.3.3', preferred=True)
-    version('2.2.1', git=git, tag='2.2.1')
+    version('develop', git=git, branch='master', clean=False)
+    version('2.3.4', git=git, tag='2.3.4', clean=False)
```

Comment:
I'm not very found of the fact we have to leave dust behind...
When fetching we could retrieve the commit_hash and save it in a fetcher variable...

Reply:
That's what we have to do when using `setuptools-scm`, too. Not very elegant, but sometimes necessary.

Reply:
Wait, this clean=False does anything else than deleting the stage dir? I meant, in GitFetchStrategy.fetch() add the line to retrieve the commit. Then the clean could happen normally (?)

Reply:
The stage directory still gets deleted. This is for source caching, where `.git` is not included in the cached tarball. When using a source mirror, this is harmful.

Reply:
> This is for source caching, where .git is not included in the cached tarball.

Would it be too bad to change that behavior?

Reply:
btw, cached source tarballs dont get named after the commit hash?

Reply:
> btw, cached source tarballs dont get named after the commit hash?

Unfortunately not. For example for `neurodamus-neocortex` there is this:
```
==> Using cached archive: /home/magkanar/spack/var/spack/cache/neurodamus-neocortex/neurodamus-neocortex-0.1.tar.gz
```

Reply:
> > This is for source caching, where .git is not included in the cached tarball.
> 
> Would it be too bad to change that behavior?

Randomly deleting things from a source checkout and not expecting things to break seems bad behavior to me :)

With `clean=False`, the SCM directories will not be removed when creating the source tarball. The caching mechanism here is fairly generic, and I think it works fine to keep the SCM directory (to conserve space, it would be more elegant to have a shallow clone). Adding a commit hash to a filename seems to ask for different kinds of trouble.

Reply:
> The stage directory still gets deleted.

I hadn't understood that... So the case is less worrying.

> To conserve space, it would be more elegant to have a shallow clone

It seems they do it:
https://github.com/BlueBrain/spack/blob/develop/lib/spack/spack/fetch_strategy.py#L674

So if we really want to tackle that 

- we could eventually include a git metadata file in the cached archive (sounds excessive)
- find a way to set make clean=False to all versions. That was the original idea but it showed to be a bit more complicated than expected

Reply:
>  * find a way to set make clean=False to all versions. That was the original idea but it showed to be a bit more complicated than expected

I'm fine with how it is currently. See also spack#9436.

Reply:
> See also [spack#9436](https://github.com/spack/spack/pull/9436).

Ah, I didn't know you introduced it! :)

Another thought. Since we are "hardcoding" the versions in source files, why not doing that before the archive phase?

Reply:
This is a good idea. I was thinking if it is possible to use something like `@run_before('archive')` and do then the replacement of the values but I don't know when exactly does the `archive` takes place and if I can refer to it somehow. I am looking at it..